File: scripts/mk_mod_schematic_recipes.rb
#!env ruby

# frozen_string_literal: true

require "nokogiri"

config_dir = "/mnt/s/Games/Steam/steamapps/common/7 Days To Die/Data/Config"
items_file = "#{config_dir}/items.xml"

raise "#{config_dir} Does not exist" unless File.exist?(config_dir) && File.directory?(config_dir)

items = Nokogiri::XML(File.read(items_file))

output = Nokogiri::XML::Builder.new do |xml|
  xml.configs do
    xml.append(xpath: "/recipes") do
      items.xpath("//item").each do |item|
        schematic_name = item.at_xpath("@name").to_s

        next unless schematic_name.match(/mod.*Schematic/)
        next unless item.at_xpath("property[@name='CreativeMode']/@value").to_s == "Player"

        xml.recipe(name: schematic_name, count: 1, craft_area: "workbench", tags: "workbenchCrafting") do
          xml.ingredient(name: item.at_xpath("property[@name='Unlocks']/@value").to_s, count: 1)
          xml.ingredient(name: "resourcePaper", count: 50)
          xml.ingredient(name: "resourceGlue", count: 5)
          xml.ingredient(name: "resourceLeather", count: 2)
        end
      end
    end
  end
end

puts output.to_xml indent: 2, save_with: Nokogiri::XML::Node::SaveOptions::DEFAULT_XML | Nokogiri::XML::Node::SaveOptions::NO_DECLARATION

File: scripts/mk_mod_schematic_recipes.py
#!/usr/bin/env python3
"""Regenerate the modschematics recipes from the game's items.xml.

Writes the recipes XML to stdout; mkall.sh redirects it into
modlets/donovan-modschematics/Config/recipes.xml.
"""

import re
import sys
from pathlib import Path

from lxml import etree

# The repo's Config symlink points at the game's Data/Config directory
config_dir = Path(__file__).resolve().parent.parent / "Config"
items_file = config_dir / "items.xml"

if not config_dir.is_dir():
    sys.exit(f"{config_dir} Does not exist")

items = etree.parse(str(items_file))

configs = etree.Element("configs")
append = etree.SubElement(configs, "append", xpath="/recipes")

for item in items.xpath("//item"):
    schematic_name = item.get("name", "")

    if not re.search(r"mod.*Schematic", schematic_name):
        continue
    if item.xpath("string(property[@name='CreativeMode']/@value)") != "Player":
        continue

    recipe = etree.SubElement(
        append, "recipe",
        name=schematic_name, count="1",
        craft_area="workbench", tags="workbenchCrafting")

    unlocks = item.xpath("string(property[@name='Unlocks']/@value)")
    etree.SubElement(recipe, "ingredient", name=unlocks, count="1")
    etree.SubElement(recipe, "ingredient", name="resourcePaper", count="50")
    etree.SubElement(recipe, "ingredient", name="resourceGlue", count="5")
    etree.SubElement(recipe, "ingredient", name="resourceLeather", count="2")

etree.indent(configs, space="  ")
print(etree.tostring(configs).decode())
